lib/ansible/executor/task_executor.py

```diff
@@ -410,6 +410,19 @@ def _run_loop(self, items):
             results.append(res)
             del task_vars[loop_var]
 
+            # clear 'connection related' plugin variables for next iteration
+            clear_plugins = {
+                'connection': self._connection._load_name,
+                'shell': self._connection._shell._load_name
+            }
+            if self._play_context.become:
+                clear_plugins['become'] = self._get_become(self._play_context.become_method)._load_name
+
+            for plugin_type, plugin_name in iteritems(clear_plugins):
+                for var in C.config.get_plugin_vars(plugin_type, plugin_name):
+                    if var in task_vars:
+                        del task_vars[var]
+
```

Comment:
soo, instead of removing them .. why don't we just 'not add them' in the first place?
https://github.com/ansible/ansible/blob/devel/lib/ansible/executor/task_executor.py#L198

we might also need to finalize things like #55219

Reply:
They're not being added in _get_loop_items, it seems like they're added in the internal execution so this may be the easiest way to clean them up.

Reply:
Yeah, they're added in `TaskExecutor._execute()`:
https://github.com/ansible/ansible/blob/9489433a4d3e2c7bcd01ec9e93684b53ec853277/lib/ansible/executor/task_executor.py#L373
->
https://github.com/ansible/ansible/blob/9489433a4d3e2c7bcd01ec9e93684b53ec853277/lib/ansible/executor/task_executor.py#L520